### PR TITLE
commander: switch to hold once takeoff is finished

### PR DESCRIPTION
**Pull request by Stifael**
switch always to hold after takeoff. this makes sense for mc, but not sure if that also applies to FW. 
@dagar?

### CODE REVIEW

**Comment by dagar** [comment]
Yes that should be fine, and effectively no change in behaviour.

**Comment by dagar** [comment]
On second thought, check the initialization of mission_result and the timing. Commander handles the state change immediately, but mission_result is only published by navigator on updates.

**Comment by dagar** [comment]
You could compare the timestamps for mission_result and commander_state. Check that they're both updated properly.

**Comment by Antiheavy** [comment]
I can't comment directly on the code.  However, I have some questions for my own understanding and some comments related to Fixed Wing:

Does it transition to Hold at/above the takeoff completion location?  Or above the takeoff starting location?  Or above the Home point (which is usually the takeoff starting location)?

Will the Hold altitude be set by the larger of MIS_LTRMIN_ALT and MIS_TAKEOFF_ALT?

This seems like a step in the right direction for improved fixed wing auto-takeoff support (I'm not talking about Mission takeoff which is pretty much what we have to use right now, and the wrong way to do it IMHO).

Does anything in this PR preclude potential future capability to "auto transition to mission after takeoff"?  https://github.com/PX4/Firmware/issues/8035

Another important piece to fixed wing auto-takeoff is this: https://github.com/PX4/Firmware/issues/6787 
  although it doesn't directly impact this PR.

**Comment by julianoes** [comment]
@dagar I just found that this will prevent auto take offs other than the first one. The reason is exactly what you wrote: the mission result is not reset and therefore re-used when it is no longer valid.

File: src/modules/commander/commander.cpp [hunk 1]
```diff
@@ -2941,9 +2941,12 @@ int commander_thread_main(int argc, char *argv[])
 			}
 		}
 
-		/* reset main state after takeoff has completed */
-		/* only switch back to posctl */
+		/* Reset main state to loiter after takeoff is completed.
+		 * Sometimes, the mission result topic is outdated and the mission is still signalled
```
Comment (Stifael) [inline src/modules/commander/commander.cpp]: signaled

**Comment by Stifael** [comment]
@Antiheavy not sure how it works for fixedwing, but for mc it is supposed to hold/hover at the position it switches from takeoff to hold

File: src/modules/commander/commander.cpp (hunk 1 above)
Comment (julianoes) [reply src/modules/commander/commander.cpp]: Thx, fixed in 45dd306fe8.

**Comment by dagar** [comment]
Added COM_TAKEOFF_ACT to configure the action after takeoff is accepted.

 - 0 : Hold (default)
 - 1 : Mission (if valid)

**Comment by julianoes** [comment]
@dagar can you explain what the purpose of this param that you added is?

As I understand it if you want to takeoff and do a mission, you just upload and do a mission and it will automatically add the required takeoff waypoint.

**Comment by dagar** [comment]
For most hand launch mission plane usage this is a simpler and safer user model.

Put plane in TAKEOFF mode and throw, once it's safely flying automatically start the mission
  vs
Sync mission with TAKEOFF waypoint, switch to to mission mode and throw. If the wind shifts do I need to change the location of the waypoint and resync? There's also the possibility of a non takeoff waypoint being selected.

Then there's the TAKEOFF waypoint itself within the mission. Altitude FOH is disabled for takeoff, so if you do things like change the mission while flying and resync the waypoint index is reset to 0 (takeoff) and the vehicle will descend at the maximum rate.

There are still lots of awkward things with FW auto takeoff and missions that need work. This is a small piece that helps.

**Comment by Antiheavy** [comment]
I’ll add that using takeoff mode vs mission takeoff waypoints simplifies and make better things like Resume Mission in QGC or other ground stations.

**Comment by julianoes** [comment]
> There are still lots of awkward things with FW auto takeoff and missions that need work. This is a small piece that helps.

I understand 😄

File: src/modules/commander/commander.cpp [hunk 2]
```diff
@@ -2941,14 +2944,18 @@ int commander_thread_main(int argc, char *argv[])
 			}
 		}
 
-		/* reset main state after takeoff has completed */
-		/* only switch back to posctl */
-		if (main_state_prev == commander_state_s::MAIN_STATE_POSCTL) {
-
-			if (internal_state.main_state == commander_state_s::MAIN_STATE_AUTO_TAKEOFF
-					&& _mission_result.finished) {
+		/* Reset main state to loiter after takeoff is completed.
```
Comment (Stifael) [inline src/modules/commander/commander.cpp]: this comment is not correct anymore.

**Comment by julianoes** [comment]
@dagar do you still have objections here, or is it ok to get in?

**Comment by dagar** [comment]
Looks good to me. I'll give it a quick test.

**Comment by julianoes** [comment]
@dagar were you able to test this?

**Comment by julianoes** [comment]
This has been tested. Why is it moved to v1.8.0?

**Comment by dagar** [comment]
I made a few minor changes to make the commander check more robust. Tested both MC and FW in SITL.

**MC**
TAKEOFF -> HOLD
https://logs.px4.io/plot_app?log=521112ce-12d0-4425-b73c-e78f5f6ece90

TAKEOFF -> MISSION
https://logs.px4.io/plot_app?log=df661ea4-e5a7-4301-92cb-33e6ca746913


**FW**
TAKEOFF -> HOLD
https://logs.px4.io/plot_app?log=3dbce25f-d81c-4d3f-91f0-7033c1d2673e

TAKEOFF -> MISSION
https://logs.px4.io/plot_app?log=5db9d0ef-2e5a-4ec9-a8e5-6cbebc1867a5


There are a few things that can be improved with fixed wing takeoff, but that doesn't need to happen here.

**Comment by hamishwillee** [comment]
@dagar So to be clear (for docs) the change is addition of COM_TAKEOFF_ACT such that after the takeoff has "completed" the vehicle will "hold" or "proceed with a mission", depending on the value of the parameter:
- Is that correct?
- By "Hold" do you mean "Hold mode" or just loiter while remaining in TAKEOFF mode?
- This is the same for MC and FW?
- I'm a bit confused above because you refer to takeoff automatically being added to missions in FW and I thought that only happened on MC?

Are there any docs changes that come from this? Specific areas where it looks possible/likely
1. Does the [Flight mode evaluation diag change](https://dev.px4.io/en/concept/flight_modes.html#flight-mode-evaluation-diagram)
2. [Takeoff mode](https://docs.px4.io/en/flight_modes/takeoff.html) - I think this is just addition of the COM_TAKEOFF_ACT for both MC and FW sections, plus maybe some words from https://github.com/PX4/Firmware/pull/8020#issuecomment-335025385 in the FW section. Does that make sense?

**Comment by dagar** [comment]
It's a mode change and it's the same for all vehicle types.

Previously in takeoff mode once completed the mode would stay the same, but the vehicle internally would be holding. Now it's explicit and the user is in a state they can click on screen in QGC to redirect.

The second part of this is the optional switch to mission mode (if a valid mission is available) after a successful takeoff (again, the mode). 

It's mainly useful for FW hand launch, but there are a number of little improvements to make in both Firmware and QGC before I'd recommend users try it.

**Comment by hamishwillee** [comment]
Thanks @dagar - I'll make some changes for this.
I'm starting to think I need to create a develop branch for docs - for all the stuff going into v1.8. Does anyone have a better idea for docs version management?

**Comment by hamishwillee** [comment]
PS More generally, really helpful if you create issues/PRs for changes that affect users/docs. I'm really happy to do the work, but very easy for me to miss changes like this.

**Pull request merged**